doc/sel.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.33.
.TH SEL "1" "February 2004" "sel 0.1.0" "User Commands"
.SH NAME
sel \- manual page for sel.  IPMI System Event Logger version 0.1.0
.SH DESCRIPTION
IPMI System Event Logger is used to view and delete SEL entries.
.SH OPTIONS
.TP
\fB\-u\fR, \fB\-\-usage\fR
Usage message
.TP
\fB\-h\fR, \fB\-\-help\fR
Show help
.TP
\fB\-V\fR, \fB\-\-version\fR
Show version
.TP
\fB\-c\fR, \fB\-\-delete-all\fR
Delete all records in SEL
.TP
\fB\-d REC-LIST\fR, \fB\-\-delete=REC-LIST\fR
Delete REC-LIST in SEL
.SH "EXAMPLES"
.B # sel 
.PP
Show all SEL entries.
.PP
.B # sel --delete=44,82
.PP
Delete entries 44 and 82.
.PP
.B # sel --delete-all
.PP
Delete all entries in SEL.
.PP
.SH "REPORTING BUGS"
Report bugs to <ab@gnu.org.in>.
.SH COPYRIGHT
Copyright \(co 2003-2004 FreeIPMI Core Team.  
This program is free software; you may redistribute it under the terms of
the GNU General Public License.  This program has absolutely no warranty.
.SH "SEE ALSO"
The full documentation for
.B sel
is maintained as a Texinfo manual.  If the
.B info
and
.B sel
programs are properly installed at your site, the command
.IP
.B info sel
.PP
should give you access to the complete manual.


NEWS
FreeIPMI 1.7.1 - XX/XX/XX
-------------------------

In 2020 the IPMI standard was frozen.  So a decision was made to try
and make the FreeIPMI 1.7.X line the last stable line in FreeIPMI.  A
number of decisions were made to remove legacy support and fix minor issues,
even at the cost of breaking backwards compatibility.

Updated Features:
-----------------

Globally
--------
o Internally allow hostnames to be up to 256 characters (previously only 64).

Libraries
---------
o libfreeipmi/sensor-read: Add error for privilege level insufficient.
o In libfreeipmi, support Intel S2600BPB OEM SEL interpretations.
o Significant refactoring of Intel OEM extensions.  Names of
  variables, macros, etc. may have changed.

Tools
-----
o In ipmi-sel, support Intel S2600BPB OEM SEL interpretations.
o In ipmi-config, support configuration of Maximum_Privilege_Cipher_Suite_Id_16.
o In ipmi-sensors, support new "altbridging" workaround for Lenovo motherboards.

Smaller Bug Fixes
-----------------

o In ipmi-config, drop support for
  Rmcpplus_Conf_Privilege:Maximum_Privilege_Cipher_Suite_Id_16.  It is
  not configurable.

o Globally, fix potential output sorting error with
  --consolidate-output.  Could alter output differently than before.

o Fix many minor output errors (i.e. typos, space alignment, lack of
  newline, etc.).  Could cause issues for those who script against
  output.

o In libfreeipmi, add missing fields for BMC Message Channel SDR
  record format.

Remove Large Amount of Legacy Support
-------------------------------------

Remove backwards compatibility, legacy output/options/fields/files, or
deprecated options.  All verified to have been gone for atleast 10
years.

For details of changes see NEWS.errata

Typo Fixes:
-----------

Fix large number of typo and minor output formattings.  They have the
potential to break against existing builds or scripts or configs that
relied on previous expected output.

See NEWS.errata for details

FreeIPMI 1.6.1 - 02/02/18
-------------------------
o Add IPv6 hostname support to FreeIPMI, all of FreeIPMI can now take
  IPv6 addresses as inputs to "host" parameters, options, or inputs.
o Support significant portions of IPMI IPv6 configuration in
  libfreeipmi.
o Add --no-session option in ipmi-raw.
o Add SDR cache options to ipmi-config.
o Legacy -f short option for --flush-cache and -Q short option for
  quiet-cache.  Backwards compatible for tools that supported it
  before.
o In ipmi-oem, support Gigabyte get-bmc-services and set-bmc-services.
o Various performance improvements:
  - Remove excessive calls to secure_memset to clear memory.
  - Remove excessive memsets and clears of data.
  - Remove unnecessary "double input checks".
  - Remove expensive input checks in libfreeipmi fiid library.
    Fallout from this may include FIID_ERR_FIELD_NOT_FOUND errors in
    different fiid functions.
  - Remove unnecessary input checks in libfreeipmi fiid library.
  - Add recent 'lookups' of fields in fiid library to internal cache.
o Various minor fixes/improvements
  - Update libfreeipmi core API to use poll() instead of select(), to
    avoid issues with applications with a high number of threads.

FreeIPMI 1.5.7 - 08/16/17
-------------------------
o In libipmimonitoring, fix several mem-leak corner cases.

FreeIPMI 1.5.6 - 07/05/17
-------------------------
o In libfreeipmi locate (used by ipmi-locate), use DMI firmware in
  sysfs if it exists.
o Minor mem-leak corner case fix in libfreeipmi.
o Minor documentation fixes.
o Minor error message clarifications.

FreeIPMI 1.5.5 - 11/02/16
-------------------------
o Fix invalid flag clear in libipmiconsole that can lead to a potential
  double close on a file descriptor.
o Support Supermicro H8SGL-F OEM sensors and events.

FreeIPMI 1.5.4 - 09/30/16
-------------------------
o Various changes/fixes in libipmiconsole
  - If user retrieves file descriptor from ipmiconsole_ctx_fd, user is
    required to close it.  ipmiconsole_ctx_destroy no longer closes
    it.  This is to avoid a potential double close which can be a
    problem for multithreaded applications.  This is a change in
    behavior, but we do not believe this will affect most applications
    since most users close the file descriptor under most scenarios
    anyways.
  - ipmiconsole_ctx_destroy() should now be called to free resources
    even if ipmiconsole_engine_teardown() has been called.  This has
    been done to create consistent behavior in the API and avoid a
    former segfault possibility.  This is a change in behavior, but we
    do not believe this will affect most applications since
    ipmiconsole_engine_teardown() is only called when an application
    is being shutdown.
  - The use of IPMICONSOLE_ENGINE_CLOSE_FD has been clarified in the
    header file.  Some of the prior text was unclear.  Behavior has
    not been changed.
o In ipmi-oem, support Intel get-bmc-services and set-bmc-services commands.
o In ipmi-oem, support Gigabyte get-nic-mode and set-nic-mode commands.
o Support Gigabyte MD90-FS0-ZB OEM SEL events.

FreeIPMI 1.5.3 - 08/01/16
-------------------------
o Update Fujitsu OEM extension product id support range.
o In ipmiconsole, support --debugfile option for more difficult
  debugging scenarios.

fish/extensions/sel.scm
;;; sel.scm: System Event Logger
;;; authors: M.P.Anand Babu <ab@gnu.org.in> 
;;; Balamurugan <bala.a@californiadigital.com>

;;; This program is free software; you can redistribute it and/or
;;; modify it under the terms of the GNU General Public License as
;;; published by the Free Software Foundation; either version 2, or (at
;;; your option) any later version.
;;; 
;;; This program is distributed in the hope that it will be useful, but
;;; WITHOUT ANY WARRANTY; without even the implied warranty of
;;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
;;; General Public License for more details.
;;; 
;;; You should have received a copy of the GNU General Public License
;;; along with this program; if not, write to the Free Software
;;; Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA
;;; 
;;; sel.scm should be automatically loaded thru init.scm

(use-modules (ice-9 getopt-long))

(define sel-exit-status 0)

(define (sel-display-usage)
  (begin 
    (display "Usage: ipmi-sel [-ic?V] [-h IPMIHOST] [-u USERNAME] [-p PASSWORD]\n")
    (display "                [-a AUTHTYPE] [-l PRIVILEGE-LEVEL] [-d REC-LIST]\n")
    (display "                [-x [FILE]] [--driver-poll-interval=USEC]\n")
    (display "                [--sms-io-base=SMS-IO-BASE] [--host=IPMIHOST]\n")
    (display "                [--username=USERNAME] [--password=PASSWORD]\n")
    (display "                [--auth-type=AUTHTYPE] [--priv-level=PRIVILEGE-LEVEL]\n")
    (display "                [--info] [--delete=REC-LIST] [--delete-all]\n")
    (display "                [--hex-dump=[FILE]] [--help] [--usage] [--version]\n")))

(define (sel-display-help)
  (begin 
    (display "Usage: ipmi-sel [OPTION...] \n")
    (display "IPMI System Event Logger is used to view and delete SEL entries.\n")
    (display "\n")
    (display "      --driver-poll-interval=USEC\n")
    (display "                             User USEC driver poll interval.\n")
    (display "      --sms-io-base=SMS-IO-BASE   SMS IO Base address.\n")
    (display "  -h, --host=IPMIHOST        Connect to IPMIHOST.\n")
    (display "  -u, --username=USERNAME    Use USERNAME instead of NULL.  Maximum USERNAME\n")
    (display "                             length is 16.\n")
    (display "  -p, --password=PASSWORD    Use PASSWORD instead of NULL.  Maximum PASSWORD\n")
    (display "                             length is 16.\n")
    (display "  -a, --auth-type=AUTHTYPE   Use AUTHTYPE instead of NONE.  Allowed values are\n")
    (display "                             NONE, MD2, MD5, PLAIN and OEM.\n")
    (display "  -l, --priv-level=PRIVILEGE-LEVEL\n")
    (display "                             Use this PRIVILEGE-LEVEL instead of USER.  Allowed\n")
    (display "                             values are CALLBACK, USER, OPERATOR, ADMIN and\n")
    (display "                             OEM.\n")
    (display "  -i, --info                 Show general information about SEL.\n")
    (display "  -d, --delete=REC-LIST      Delete given SEL records entry.\n")
    (display "  -c, --delete-all           Delete all SEL entries.\n")
    (display "  -x, --hex-dump[=FILE]      Hex-dump SEL entries optionally to FILE.\n")
    (display "  -?, --help                 Give this help list.\n")
    (display "      --usage                Give a short usage message.\n")
    (display "  -V, --version              Print program version.\n")
    (display "\n")
    (display "Mandatory or optional arguments to long options are also mandatory or optional\n")
    (display "for any corresponding short options.\n")
    (display "\n")
    (display "Report bugs to <freeipmi-devel@gnu.org>.\n")))

(define (sel-display-version)
  (begin 
    (display (string-append "IPMI System Event Logger [ipmi-sel-" (fi-version) "]\n"))
    (display "Copyright (C) 2003-2005 FreeIPMI Core Team\n")
    (display "This program is free software; you may redistribute it under the terms of\n")
    (display "the GNU General Public License.  This program has absolutely no warranty.\n")))

(define (sel-argp args)
  (catch 'misc-error 
	 (lambda () 
	   (let* ((sel-cmd-args '())
		  (option-spec '((driver-poll-interval (single-char #\203) (value #t))
				 (sms-io-base   (single-char #\204) (value #t))
				 (host          (single-char #\h)   (value #t))
				 (username      (single-char #\u)   (value #t))
				 (password      (single-char #\p)   (value #t))
				 (auth-type     (single-char #\a)   (value #t))
				 (priv-level    (single-char #\l)   (value #t))
				 (help          (single-char #\?)   (value #f))
				 (usage         (single-char #\377) (value #f))
				 (version       (single-char #\V)   (value #f))
				 (info          (single-char #\i)   (value #f))
				 (delete        (single-char #\d)   (value #t))
				 (delete-all    (single-char #\c)   (value #f))
				 (hex-dump      (single-char #\x)   (value optional))
				 (delete-range  (single-char #\r)   (value #t))))
				 ;;(delete-event-id (single-char #\e) (value #t))))
		  (options (getopt-long args option-spec))
		  (poll-interval  (option-ref options 'poll-interval #f))
		  (sms-io-base    (option-ref options 'sms-io-base   #f))
		  (host           (option-ref options 'host          #f))
		  (username       (option-ref options 'username      #f))
		  (password       (option-ref options 'password      #f))
		  (auth-type      (option-ref options 'auth-type     #f))
		  (priv-level     (option-ref options 'priv-level    #f))
		  (help-wanted    (option-ref options 'help          #f))
		  (usage-wanted   (option-ref options 'usage         #f))
		  (version-wanted (option-ref options 'version       #f))
		  (info-wanted    (option-ref options 'info          #f))
		  (delete-list    (option-ref options 'delete        #f))
		  (delete-all     (option-ref options 'delete-all    #f))
		  (hex-dump-name  (option-ref options 'hex-dump      #f))
		  (delete-range   (option-ref options 'delete-range  #f))
		  (delete-event-id (option-ref options 'delete-event-id #f))
		  (extra-args     (option-ref options '()            #f)))
	     ;; extra arguments
	     (if (and (not (null? extra-args)) (list? sel-cmd-args))
		 (begin 
		   (display "Usage: ipmi-sel [OPTION...] \n"
			    (current-error-port))
		   (display "Try `ipmi-sel --help' or `ipmi-sel --usage' for more information.\n"
			    (current-error-port))
		   (set! sel-exit-status 64)
		   (set! sel-cmd-args #f)))
	     ;; --driver-poll-interval (0)
	     (if (and (string? poll-interval) (list? sel-cmd-args))
		 (begin 
		   (set! poll-interval (string->number poll-interval))
		   (if (boolean? poll-interval)
		       (begin 
			 (display "Usage: ipmi-sel [OPTION...] \n"
				  (current-error-port))
			 (display "Try `ipmi-sel --help' or `ipmi-sel --usage' for more information.\n"
				  (current-error-port))
			 (set! sel-exit-status 64)
			 (set! sel-cmd-args #f)))))
	     (if (list? sel-cmd-args)
		 (set! sel-cmd-args (append sel-cmd-args 
					    (list poll-interval))))
	     ;; --sms-io-base (1)
	     (if (and (string? sms-io-base) (list? sel-cmd-args))
		 (begin 
		   (set! sms-io-base (string->number sms-io-base))
		   (if (boolean? sms-io-base)
		       (begin 
			 (display "Usage: ipmi-sel [OPTION...] \n"
				  (current-error-port))
			 (display "Try `ipmi-sel --help' or `ipmi-sel --usage' for more information.\n"
				  (current-error-port))
			 (set! sel-exit-status 64)
			 (set! sel-cmd-args #f)))))
	     (if (list? sel-cmd-args)
		 (set! sel-cmd-args (append sel-cmd-args 
					    (list sms-io-base))))
	     ;; --host (2)
	     (if (list? sel-cmd-args)
		 (set! sel-cmd-args (append sel-cmd-args 
					    (list host))))
	     ;; --username (3)
	     (if (and (string? username) (list? sel-cmd-args))
		 (begin 
		   (if (> (string-length username) 16)
		       (begin 
			 (display "Usage: ipmi-sel [OPTION...] \n"
				  (current-error-port))
			 (display "Try `ipmi-sel --help' or `ipmi-sel --usage' for more information.\n"
				  (current-error-port))
			 (set! sel-exit-status 64)
			 (set! sel-cmd-args #f)))))
	     (if (list? sel-cmd-args)
		 (set! sel-cmd-args (append sel-cmd-args 
					    (list username))))
	     ;; --password (4)
	     (if (and (string? password) (list? sel-cmd-args))
		 (begin 
		   (if (> (string-length password) 16)
		       (begin 
			 (display "Usage: ipmi-sel [OPTION...] \n"
				  (current-error-port))
			 (display "Try `ipmi-sel --help' or `ipmi-sel --usage' for more information.\n"
				  (current-error-port))
			 (set! sel-exit-status 64)
			 (set! sel-cmd-args #f)))))
	     (if (list? sel-cmd-args)
		 (set! sel-cmd-args (append sel-cmd-args 
					    (list password))))
	     ;; --auth-type (5)
	     (if (and (string? auth-type) (list? sel-cmd-args))
		 (cond 
		  ((string-ci=? auth-type "none")
		   (set! auth-type 0))
		  ((string-ci=? auth-type "md2")
		   (set! auth-type 1))
		  ((string-ci=? auth-type "md5")
		   (set! auth-type 2))
		  ((string-ci=? auth-type "plain")
		   (set! auth-type 4))
		  ((string-ci=? auth-type "oem")
		   (set! auth-type 5))
		  (else 
		   (begin 
		     (display "Usage: ipmi-sel [OPTION...] \n"
			      (current-error-port))
		     (display "Try `ipmi-sel --help' or `ipmi-sel --usage' for more information.\n"
			      (current-error-port))
		     (set! sel-exit-status 64)
		     (set! sel-cmd-args #f))))
		 (set! auth-type 0))
	     (if (list? sel-cmd-args)
		 (set! sel-cmd-args (append sel-cmd-args 
					    (list auth-type))))
	     ;; --priv-level (6)
	     (if (and (string? priv-level) (list? sel-cmd-args))
		 (cond 
		  ((string-ci=? priv-level "callback")
		   (set! priv-level 1))
		  ((string-ci=? priv-level "user")
		   (set! priv-level 2))
		  ((string-ci=? priv-level "operator")
		   (set! priv-level 3))
		  ((string-ci=? priv-level "admin")
		   (set! priv-level 4))
		  ((string-ci=? priv-level "oem")
		   (set! priv-level 5))
		  (else 
		   (begin 
		     (display "Usage: ipmi-sel [OPTION...] \n"
			      (current-error-port))
		     (display "Try `ipmi-sel --help' or `ipmi-sel --usage' for more information.\n"
			      (current-error-port))
		     (set! sel-exit-status 64)
		     (set! sel-cmd-args #f))))
		 (set! priv-level 2))
	     (if (list? sel-cmd-args)
		 (set! sel-cmd-args (append sel-cmd-args 
					    (list priv-level))))
	     ;; --help (7)
	     (if (list? sel-cmd-args)
		 (set! sel-cmd-args (append sel-cmd-args 
					    (list help-wanted))))
	     ;; --usage (8)
	     (if (list? sel-cmd-args)
		 (set! sel-cmd-args (append sel-cmd-args 
					    (list usage-wanted))))
	     ;; --version (9)
	     (if (list? sel-cmd-args)
		 (set! sel-cmd-args (append sel-cmd-args 
					    (list version-wanted))))
	     ;; --info (10) SEL specific
	     (if (list? sel-cmd-args)
		 (set! sel-cmd-args (append sel-cmd-args 
					    (list info-wanted))))
	     ;; --delete-list (11) SEL specific
	     (if (and (string? delete-list) (list? sel-cmd-args))
		 (begin 
		   (set! delete-list (sentence->tokens (string-replace 
							delete-list 
							#\, #\space)))
		   (if (or (list? (member #f (map number? delete-list)))
			   (null? delete-list))
		       (begin 
			 (display "Usage: ipmi-sel [OPTION...] \n"
				  (current-error-port))
			 (display "Try `ipmi-sel --help' or `ipmi-sel --usage' for more information.\n"
				  (current-error-port))
			 (set! sel-exit-status 64)
			 (set! sel-cmd-args #f)))))
	     (if (list? sel-cmd-args)
		 (set! sel-cmd-args (append sel-cmd-args 
					    (list delete-list))))
	     ;; --delete-all (12) SEL specific
	     (if (list? sel-cmd-args)
		 (set! sel-cmd-args (append sel-cmd-args 
					    (list delete-all))))
	     ;; --hex-dump-name (13) SEL specific
	     (if (list? sel-cmd-args)
		 (set! sel-cmd-args (append sel-cmd-args 
					    (list hex-dump-name))))

	     ;; --delete-range (14) SEL specific
	     (if (and (string? delete-range) (list? sel-cmd-args))
		 (begin 
		   (set! delete-range (sentence->tokens (string-replace 
							delete-range 
							#\- #\space)))
		   (if (or (list? (member #f (map number? delete-range)))
			   (null? delete-range)
			   (not (= (length delete-range) 2))
			   (> (car delete-range) (cadr delete-range)))
		       (begin 
			 (display "Usage: ipmi-sel [OPTION...] \n"
				  (current-error-port))
			 (display "Try `ipmi-sel --help' or `ipmi-sel --usage' for more information.\n"
				  (current-error-port))
			 (set! sel-exit-status 64)
			 (set! sel-cmd-args #f)))))
	     (if (list? sel-cmd-args)
		 (set! sel-cmd-args (append sel-cmd-args 
					    (list delete-range))))

	     ; --delete-event-id ID (15) SEL specific
	     (if (and (string? delete-event-id)
		      (list? sel-cmd-args)
		      (not (string->number delete-event-id)))
		 (begin
		   (display "Usage: ipmi-sel [OPTION...] \n"
			    (current-error-port))
		   (display "Try `ipmi-sel --help' or `ipmi-sel --usage' for more information.\n"
			    (current-error-port))
		   (set! sel-exit-status 64)
		   (set! sel-cmd-args #f)))
	     (if (list? sel-cmd-args)
		 (begin
		   (and (string? delete-event-id)
			(set! delete-event-id (string->number
					       delete-event-id)))
		   (set! sel-cmd-args (append sel-cmd-args
					    (list delete-event-id)))))

	   sel-cmd-args))
	 (lambda (k args . opts)
	   (display "sel: error: " (current-error-port))
	   (display (cadr opts) (current-error-port))
	   (display "\n" (current-error-port))
	   (display "Usage: ipmi-sel [OPTION...] \n"
		    (current-error-port))
	   (display "Try `ipmi-sel --help' or `ipmi-sel --usage' for more information.\n"
		    (current-error-port))
	   (set! sel-exit-status 64)
	   #f)))

(define (sel-get-help-option cmd-args)
  (list-ref cmd-args 7))

(define (sel-get-usage-option cmd-args)
  (list-ref cmd-args 8))

(define (sel-get-version-option cmd-args)
  (list-ref cmd-args 9))

(define (sel-get-info-option cmd-args)
  (list-ref cmd-args 10))

(define (sel-get-delete-list-option cmd-args)
  (list-ref cmd-args 11))

(define (sel-get-delete-all-option cmd-args)
  (list-ref cmd-args 12))

(define (sel-get-hex-dump-option cmd-args)
  (list-ref cmd-args 13))

(define (sel-get-delete-range-option cmd-args)
  (list-ref cmd-args 14))

(define (sel-get-delete-event-id-option cmd-args)
  (list-ref cmd-args 15))

(define (sel-display-entry sel)
  (display (list-ref sel 0)) (display ":")
  (display (list-ref sel 1)) (display ":")
  (display (list-ref sel 2)) (display ":")
  (display (list-ref sel 3)) (newline)
  (force-output))

(define (sel-display-all-entry)
  (letrec 
      ((sel-display-entry 
	(lambda (sel)
	  (if (not (null? sel))
	      (begin 
		(display (list-ref sel 0))
		(and (string? (list-ref sel 1)) 
		     (begin (display ":") (display (list-ref sel 1))))
		(and (string? (list-ref sel 2)) 
		     (begin (display ":") (display (list-ref sel 2))))
		(and (string? (list-ref sel 3)) 
		     (begin (display ":") (display (list-ref sel 3))))
		(and (string? (list-ref sel 4)) 
		     (begin (display ":") (display (list-ref sel 4))))
		(and (string? (list-ref sel 5)) 
		     (begin (display ":") (display (list-ref sel 5))))
		(newline)
		;(display (strftime "%d-%b-%Y %H:%M:%S" (localtime (list-ref sel 1)))) 
		(force-output)
		(sel-display-entry (fi-sel-get-next-entry)))))))
    (sel-display-entry (fi-sel-get-first-entry))))

(define (sel-get-record-list)
  (letrec
      ((remaining-records (lambda ()
			    (let ((entry (fi-sel-get-next-entry-raw)))
			      (cond ((null? entry) '())
				    (else (append (list entry)
						  (remaining-records))))))))
    (let ((first-entry (fi-sel-get-first-entry-raw)))
      (append (list (if (null? first-entry) '() first-entry))
	  (if (null? first-entry) '() (remaining-records))))))

(define (sel-record-foreach call-me)
  (map call-me (sel-get-record-list)))

(define (sel-delete-record-list delete-list)
  (if (not (null? delete-list))
      (begin 
	(if (not (fi-sel-delete-entry (car delete-list)))
	    (set! sel-exit-status 1))
	(sel-delete-record-list (cdr delete-list)))))

(define (sel-hex-dump)
  (let ((info (fi-sel-get-info-binary)))
    (sel-display-info gmtime info))
  (let loop ((first-entry (fi-sel-get-first-entry-hex)))
    (if (string? first-entry)
        (begin
          (display first-entry)
          (loop (fi-sel-get-next-entry-hex))))))

(define (sel-display-info time-splitter info)
  (if info
      (let ((last-add-time-string
             (strftime "%m/%d/%Y - %H:%M:%S" (time-splitter (list-ref info 3))))
            (last-erase-time-string
             (strftime "%m/%d/%Y - %H:%M:%S" (time-splitter (list-ref info 4)))))
        (simple-format
         #t (string-append
             "Version                     IPMI v~A.~A\n"
             "Number of Entries           ~A\n"
             "Last Add Time               ~A\n"
             "Last Erase Time             ~A\n"
             "Free Space Remaining        ~A\n\n")
         (list-ref info 0) (list-ref info 1)
         (list-ref info 2)
         last-add-time-string
         last-erase-time-string
         (list-ref info 5)))))

(define (sel-display-flags info)
  (if info
      (begin
        (display "Overflow Flag                                     ")
        (display (if (list-ref info 6) "Yes\n" "No\n"))
        (display "Delete SEL Command Supported                      ")
        (display (if (list-ref info 7) "Yes\n" "No\n"))
        (display "Partial Add SEL Enty Command Supported            ")
        (display (if (list-ref info 8) "Yes\n" "No\n"))
        (display "Reserve SEL Command Supported                     ")
        (display (if (list-ref info 9) "Yes\n" "No\n"))
        (display "Get SEL Allocation Information Command Supported  ")
        (display (if (list-ref info 10) "Yes\n" "No\n")))))

(define (range->list tuple_list)
  (cond ((= (car tuple_list) (cadr tuple_list)) (cdr tuple_list))
        (else (append (list (car tuple_list))
                      (range->list (list (+ 1 (car tuple_list))
                                         (cadr tuple_list)))))))

(define (sel-main cmd-args)
  (cond 
   ((sel-get-help-option cmd-args)
    (sel-display-help))
   ((sel-get-usage-option cmd-args)
    (sel-display-usage))
   ((sel-get-version-option cmd-args)
    (sel-display-version))
   (else 
    (and (fi-ipmi-open cmd-args)
	 (begin 
	   (cond 
	    ((sel-get-info-option cmd-args)
	     (let ((info (fi-sel-get-info-binary)))
	       (sel-display-info localtime info)
	       (sel-display-flags info)))
	    ((sel-get-hex-dump-option cmd-args)
	     (sel-hex-dump))
	    ((string? (sel-get-hex-dump-option cmd-args))
	     (with-output-to-file (sel-get-hex-dump-option cmd-args) 
	       sel-hex-dump))
	    ((sel-get-delete-all-option cmd-args)
	     (fi-sel-clear))
	    ((sel-get-delete-list-option cmd-args)
	     (sel-delete-record-list (sel-get-delete-list-option cmd-args)))
            ((sel-get-delete-range-option cmd-args)
             (sel-delete-record-list (range->list 
				      (sel-get-delete-range-option cmd-args))))
	    ((sel-get-delete-event-id-option cmd-args)
	     (sel-record-foreach (lambda (a-sel)
				   (and (= (length a-sel) 16)
					(= (list-ref a-sel 11) 
					   (sel-get-delete-event-id-option 
					    cmd-args))
					(fi-sel-delete-entry 
					 (+ (list-ref a-sel 0)
					    (* 256 (list-ref a-sel 1))))))))
	    (else 
	     (sel-display-all-entry)))
	   (fi-ipmi-close))))))

(define (sel args)
  "fish sel main"
  (let ((cmd-args (sel-argp (append (list "sel") 
				    (list->strlist args)))))
    (if (list? cmd-args)
	(sel-main cmd-args))))

(fi-register-command! 
 (list "sel" 
       (string-append 
	"Usage: sel [--driver-poll-interval=USEC] [--sms-io-base=SMS-IO-BASE]\n"
	"           [--host=IPMIHOST] [--username=USERNAME]\n"
	"           [--password=PASSWORD] [--auth-type=AUTHTYPE]\n"
	"           [--priv-level=PRIVILEGE-LEVEL] [--info] [--delete=REC-LIST]\n"
	"           [--delete-all] [--delete-range=START-END] [--hex-dump=[FILE]]\n"
        "           [--help] [--usage] [--version]\n"
	"\n"
	"          Used to view and delete SEL entries.")))
